pkg/channel/distributed/common/kafka/sarama/sarama.go

```diff
@@ -308,73 +173,64 @@ func extractRootCerts(saramaConfigYamlString string) (string, *x509.CertPool, er
+// them in as the "ignore" parameter.
+func ConfigEqual(config1, config2 *sarama.Config, ignore ...interface{}) bool {
+	// If some of the types in the sarama.Config struct are not ignored, these kinds of errors will appear:
+	// panic: cannot handle unexported field at {*sarama.Config}.Consumer.Group.Rebalance.Strategy.(*sarama.balanceStrategy).name
 
-	// We Always Want Success Messages From Producer
-	config.Producer.Return.Successes = true
-}
+	// Note that using the types directly from config1 is convenient (it allows us to call IgnoreTypes instead of the
+	// more complicated IgnoreInterfaces), but it will fail if, for example, config1.Consumer.Group.Rebalance is nil
 
-// Utility Function For Configuring Common Settings For Admin/Producer/Consumer
-func UpdateSaramaConfig(config *sarama.Config, clientId string, username string, password string) {
+	// However, the sarama.NewConfig() function sets all of these values to a non-nil default, so the risk
+	// is minimal and should be caught by one of the several unit tests for this function if the sarama vendor
+	// code is updated and these defaults become something invalid at that time)
 
-	// Set The ClientID For Logging
-	config.ClientID = clientId
+	ignoreTypeList := append([]interface{}{
+		config1.Consumer.Group.Rebalance.Strategy,
+		config1.MetricRegistry,
+		config1.Producer.Partitioner},
+		ignore...)
+	ignoredTypes := cmpopts.IgnoreTypes(ignoreTypeList...)
 
-	// Set The SASL Username / Password
-	config.Net.SASL.User = username
-	config.Net.SASL.Password = password
+	// If some interfaces are not included in the "IgnoreUnexported" list, these kinds of errors will appear:
+	// panic: cannot handle unexported field at {*sarama.Config}.Net.TLS.Config.mutex: "crypto/tls".Config
 
-	// Force Required Configuration
-	enforceSaramaConfig(config)
-}
+	// Note that x509.CertPool and tls.Config are created here explicitly because config1/config2 may not
+	// have those fields, and results in a nil pointer panic if used in the IgnoreUnexported list indirectly
+	// like config1.Version is (Version is required to be present in a sarama.Config struct).
 
-// This custom marshaller converts a sarama.Config struct to our CompareSaramaConfig using the ability of go
-// to initialize one struct with another identical one (aside from field tags).  It then uses the CompareSaramaConfig
-// field tags to skip the fields that cause problematic JSON output (such as function pointers).
-// An error of "cannot convert *config (type sarama.Config) to type CompareSaramaConfig" indicates that the
-// CompareSaramaConfig structure above needs to be brought in sync with the current config.Sarama struct.
-func MarshalSaramaJSON(config *sarama.Config) (string, error) {
-	bytes, err := json.Marshal(CompareSaramaConfig(*config))
-	if err != nil {
-		return "", err
-	}
-	return string(bytes), nil
-}
+	ignoredUnexported := cmpopts.IgnoreUnexported(config1.Version, x509.CertPool{}, tls.Config{})
 
```

Comment:
Would be awesome if we could go another level deep on the `tls.Config{}` exclusion.  We have an immediate need to be able to specify some of the bools in tls.Config from the configmap.  I know it's a pain (lot to exclude) but otherwise we have more custom extraction logic similar to Version to deal with.   Thoughts?

Reply:
The IgnoreUnexported call specifies structures which need to have private fields ignored in the comparison.  For tls.Config this includes only mutex, sessionTicketKeys, and autoSessionTicketKeys.  InsecureSkipVerify is public and as such should still be comparable; I've added calls in the TestMergeSaramaSettings() and TestSaramaConfigEqual() functions to explicitly check this particular action for peace of mind.

Reply:
perfect - thanks Eric!